src/pool/srv_pool.c

```diff
@@ -3879,6 +4024,188 @@ ds_pool_update(uuid_t pool_uuid, crt_opcode_t opc,
 	return rc;
 }
 
+static int
+ds_pool_extend_internal(struct rdb_tx *tx, struct pool_svc *svc,
+		uint32_t nnodes, uuid_t target_uuids[],
+		d_rank_list_t *rank_list, uint32_t ndomains,
+		int32_t *domains, bool *updated_p, uint32_t *map_version_p,
+		struct rsvc_hint *hint)
+{
+	struct pool_buf		*map_buf;
+	struct pool_map		*map = NULL;
+	uint32_t 		map_version;
+	uuid_t			*uuids = NULL;
+	bool			updated = false;
+	int			ntargets;
+	int			rc;
+
+	ntargets = nnodes * dss_tgt_nr;
+
+	/* Create a temporary pool map based on the last committed version. */
+	rc = read_map(tx, &svc->ps_root, &map);
+	if (rc != 0)
+		return rc;
+
+	map_version = pool_map_get_version(map) + 1;
+
+	/* Prepare the pool map attribute buffers. */
+	map_buf = pool_buf_alloc(ndomains + nnodes + ntargets);
+	if (map_buf == NULL)
+		return -DER_NOMEM;
+
+	/* Make a sorted target UUID array to determine target IDs. */
+	D_ALLOC_ARRAY(uuids, nnodes);
+	if (uuids == NULL)
+		D_GOTO(out_map_buf, rc = -DER_NOMEM);
+	memcpy(uuids, target_uuids, sizeof(uuid_t) * nnodes);
+	qsort(uuids, nnodes, sizeof(uuid_t), uuid_compare_cb);
+
+	/* Fill the pool_buf out. */
+	rc = pool_buf_fill_extend(map, map_buf, nnodes, map_version, ndomains,
+			    domains, target_uuids, rank_list, uuids, &updated);
+	if (rc != 0)
+		D_GOTO(out_uuids, rc);
+	if(!updated)
+		D_GOTO(out_map_buf, rc);
+
+	/* Extend the current pool map */
+	rc = pool_map_extend(map, map_version, map_buf);
+	if(rc != 0)
+		D_GOTO(out_map, rc);
+
+ 	/* Write the new pool map. */
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs